storage/simulation/cluster.go

```diff
@@ -104,29 +113,102 @@ func (c *Cluster) addStore(nodeID proto.NodeID) *Store {
-	newRng := newRange(rangeID)
+	newRng := newRange(rangeID, c.allocator)
 	c.ranges[rangeID] = newRng
 	return newRng
 }
 
+// splitRangeRandom splits a random range from within the cluster.
 func (c *Cluster) splitRangeRandom() {
 	rangeID := proto.RangeID(c.rand.Int63n(int64(len(c.ranges))))
 	c.splitRange(rangeID)
 }
 
+// splitRangeLast splits the last added range in the cluster.
 func (c *Cluster) splitRangeLast() {
 	rangeID := proto.RangeID(len(c.ranges) - 1)
 	c.splitRange(rangeID)
 }
 
+// splitRange "splits" a range. This split creates a new range with new
+// replicas on the same stores as the passed in range. The new range has the
+// same zone config as the original range.
 func (c *Cluster) splitRange(rangeID proto.RangeID) {
 	newRange := c.addRange()
 	originalRange := c.ranges[rangeID]
 	newRange.splitRange(originalRange)
 }
 
+// runEpoch steps through a single instance of the simulator. Each epoch
+// performs the following steps:
+// 1) The status of every store is gossiped so the store pool is up to date.
+// 2) Each replica on every range calls the allocator to determine if there are
+//    any actions required.
+// 3) The replica on each range with the highest priority executes it's action.
+// 4) The current status of the cluster is output.
+func (c *Cluster) runEpoch() {
+	c.epoch++
+
+	// Gossip all the store updates.
+	c.gossipStores()
+
+	// Determine next operations for all ranges.
+	for _, r := range c.ranges {
+		r.prepareActions()
+	}
+
+	// Execute the determined operations.
+	c.performActions()
+
+	// Output the update.
+	fmt.Println(c.StringEpoch())
+}
+
+// gossipStores gossips all the most recent status for all stores.
+func (c *Cluster) gossipStores() {
+	storesRangeCounts := make(map[proto.StoreID]int)
+	for _, r := range c.ranges {
+		for _, storeID := range r.getStoreIDs() {
+			storesRangeCounts[storeID]++
+		}
+	}
+
+	c.storeGossiper.GossipWithFunction(c.storeIDs, func() {
+		for storeID, store := range c.stores {
+			if err := store.gossipStore(storesRangeCounts[storeID]); err != nil {
+				fmt.Printf("Error gossiping store %d: %s\n", storeID, err)
+			}
+		}
+	})
+}
+
+// performActions performs a single action, if required, for each range.
+func (c *Cluster) performActions() {
+	for rangeID, r := range c.ranges {
+		nextAction := r.getNextAction()
+		switch nextAction.action {
+		case storage.AAAdd:
+			newStoreID := r.getAllocateTarget()
+			r.addReplica(c.stores[newStoreID])
+		case storage.AARemoveDead:
+			// TODO(bram): implement this.
+			fmt.Printf("Range %d - Repair\n", rangeID)
+		case storage.AARemove:
+			// TODO(bram): implement this.
+			fmt.Printf("Range %d - Remove\n", rangeID)
+		case storage.AANoop:
+			if nextAction.rebalance {
+				// TODO(bram): implement this.
+				fmt.Printf("Range %d - Rebalance\n", rangeID)
+			}
+		}
+	}
+}
+
 // String prints out the current status of the cluster.
 func (c *Cluster) String() string {
+	var buf bytes.Buffer
```

Comment:
The next line prints the cluster info.  I wanted to logically split it up.
- Cluster Info
- Node Info
- Store Info
- Range Info